containers - complete docker-compose to run the entire system

ilausuch opened the pull request:
- Add the client.conf and set the environment vars OPENQA_SCHEDULER_HOST
  and OPENQA_WEB_SOCKETS_HOST to fix the webui-scheduler communication
- Add a worker service as part of the docker-compose
- Add the worker config
- Add wait loops for an organized boot: First webui (initializes the DB),
  then the rest web UI and the worker

https://progress.opensuse.org/issues/76978

okurz reviewed (commented):
You are now changing the already existing docker compose definition but without an automatic test the functionality can not be ensured which I would suggest to do first. This should be doable within circleCI

ilausuch commented on container/webui/conf/openqa.ini:
```diff
@@ -1,5 +1,5 @@
 [auth]
-method = OpenID
+method = Fake
```
Well, the reason to use Fake is that if we use OpenID we will had to add more complexity. The OpenID gets as validation url the webui (http://webui instead of http://localhost:9526) And of course the token is not valid for the good url
```
naive_verify_failed_return: Direct contact invalidated ID provider response.
```
It could be solve creating a entry in /etc/hosts but forces the user to do this. Additionally the api_key should be created as in the bootstrap script. But I am not sure if will be valid before login.

On the contrary with Fake, the current docker-compose specification works.

In my opinion, the complexity should be moved to production deployments.

Martchus replied:
As long as the instance is only accessible from the loopback device using fake auth should be fine. However, the problem of the wrong URL could likely be fixed by setting the `base_url` within the `[global]` section to the desired URL.

kalikiana replied:
> As long as the instance is only accessible from the loopback device using fake auth should be fine. However, the problem of the wrong URL could likely be fixed by setting the `base_url` within the `[global]` section to the desired URL.

I'd say try out if this works, assuming it takes a few minutes or so. If not, record the error - or if you tried it before, mention it here again.

I think we may have discussed this part but can't find a record of it. I'd consider proper auth a nice to have that can be a follow-up. Same category as TLS imo. I'd really like it, but it's not necessary for the quick setup story.

ilausuch replied:
I tried with OpenID as @Martchus said and works. And also, the default api_key exists from the begining. So keeping with OpenID.

ilausuch replied:
But I realized that I don't have the administration options 
![image](https://user-images.githubusercontent.com/11134265/109670268-3b395280-7b73-11eb-837e-e46ceb6b3e1b.png)

Martchus replied:
Right, annoyingly the user (I literally mean a row in our users table) which is created on OpenID login won't have operator/admin permissions set.

You *could* run `ALTER TABLE users ALTER COLUMN is_operator SET DEFAULT 1;` and `ALTER TABLE users ALTER COLUMN is_admin SET DEFAULT 1;` to make every new user operator/admin. However, that would obviously mean everybody who can login can do everything which is likely also not what we want.

Not sure yet how to solve this in an elegant way.

kalikiana replied:
The very first user created is always admin, regardless of auth method, no?

ilausuch replied:
What do you (others) think? 
I propose to keep OpenID and comment in the commentary that is possible to change the auth to Fake to have full functionality. I don't know if is a good idea add more documentation with the SQL query. What do you think?

ilausuch replied:
False assertion. With OpenID we don't have any apikey until login, so worker cannot initizalizate. The reason of before was working was because I was keeping the DB from when I used Fake
So To not do this more complex, I am going to keep it with Fake

kalikiana replied:
You need to call the `/login` route once, yes, to create the first admin user

Martchus replied:
> The very first user created is always admin, regardless of auth method, no?

Judging by the code that's not the case. For fake auth the permissions are hard-coded (so "Demo" is always admin) and for OpenID new users are never admin/operator.

And @ilausuch is also right about the API key/secret. When relying on OpenID we needed to generate the key for that user somehow *after* the first OpenID login.

Martchus commented on container/webui/nginx.conf:
```diff
@@ -14,17 +14,18 @@ server {
       proxy_set_header Connection "upgrade";
       proxy_set_header Host $host;
-
+      proxy_read_timeout     300;
+      proxy_connect_timeout  300;
```
This might make sense for the liveviewhandler as well although it should usually try to re-connect when running into a timeout.

ilausuch replied:
done

Martchus reviewed (changes requested):
Besides the OpenID point it looks generally good (as far as I can tell from my limited knowledge about docker-compose). Could you check whether the documentation found within `container/README.md` needs to be extended or updated?

The commit message should also be reformatted to match our style, e.g. `containers: Extend docker-compose to run the entire system`.

kalikiana commented:
> You are now changing the already existing docker compose definition but without an automatic test the functionality can not be ensured which I would suggest to do first. This should be doable within circleCI

Discussing this with @ilausuch in chat atm (with some technicaly difficulties). I would lean towards doing this in a separate PR since it's non-trivial.

ilausuch commented:
I added the documentation and also the option to change the images for containers

Martchus commented on container/README.md:
```diff
@@ -13,20 +13,141 @@ If a section or action is tagged as "Fedora", it should be applied only for a
+
+1. First run the docker-compose up
+2. Then enter to the web UI page and do Login
+3. A new api/key is created. Check it out in the db container
+
+        $ docker-compose exec db psql -U openqa openqa
+        openqa=# SELECT * FROM api_keys;
```
I like to have documentation for accessing the database within the container. However, I'm wondering whether this is really necessary for simply viewing the API credentials at this point. Wouldn't it be possible to simply show them within the web UI?

ilausuch replied:
Yes, you're right, better the UI

Martchus commented on container/README.md:
```diff
@@ -13,20 +13,141 @@ If a section or action is tagged as "Fedora", it should be applied only for a
+
+        $ docker-compose exec db psql -U openqa openqa
+        openqa=# SELECT * FROM api_keys;
+4. Modify the file client.conf and change all the api/keys
```
```suggestion
4. Modify the file `client.conf` and change **all** the api/keys
```

Martchus commented on container/README.md:
````diff
@@ -13,20 +13,141 @@ If a section or action is tagged as "Fedora", it should be applied only for a
+1. To expose the SSL port uncomment this line in the docker-compose.yaml file
+   in the service nginx.
+
+```
````
It would be nice to fix the indentation so the additional paragraphs within the list items actually belong to them.

ilausuch commented:
Thanks @Martchus for your review. Still remains some changes

ilausuch commented:
Recent changes
- I added the missing worker health-check
- Now the number of workers are controlled by an environment variable
- I reflect this last point in the documentation and I removed the part that is obsolete

kalikiana commented:
Since our conversation in chat was cut short, for the record:

- worker doesn't seem to work as-is
- images built from git and provided via env variables work
- do we know if 1) `qemu-tools` is installed and 2) `kvm` and `_openqa-worker` groups were added in the image? these were the fixes from #3715
- maybe [poo#88754](https://progress.opensuse.org/issues/88754#note-2) is related

ilausuch commented:
I pushed a new commit that fixes a problem with the factory and testresults directories. These has to be accessible by gekotest user

ilausuch commented:
Since last commit I detected that the live view is not working

Martchus commented:
> Maybe this makes sense to leave for a follow-up PR.

I suppose re-needling and local test development/investigation are use-cases we should cover. Considering that providing a half-broken setup will leave a bad first impression I don't think postponing this is a good idea.

Martchus commented:
> Since last commit I detected that the live view is not working

@ilausuch "Not working" is not much to go on so I'll just throw the generic troubleshooting guide at you :-)

https://open.qa/docs/#debugdevelmode

okurz commented:
please rebase after https://github.com/os-autoinst/openQA/pull/3772 merged
